kernel/inductive.ml

```diff
@@ -178,11 +185,39 @@ exception SingletonInductiveBecomesProp of Id.t
 
 let subst_univs_sort subs = function
 | Sorts.Prop | Sorts.Set | Sorts.SProp as s -> s
-| Sorts.Type u -> Sorts.sort_of_univ (Univ.subst_univs_universe subs u)
+| Sorts.Type u ->
+  (* We implement by hand a max on universes that handles Prop *)
+  let u = Universe.repr u in
+  let supern u n = iterate Universe.super n u in
+  let map (u, n) =
+    if Level.is_prop u || Level.is_sprop u then assert false
+    else if Level.is_set u then [Universe.type0, n]
+    else match Level.Map.find u subs with
+    | [] ->
+      if Int.equal n 0 then
+        (* This is an instantiation of a template universe by Prop, ignore it *)
+        []
+      else
+        (* Prop + S n actually means Set + S n *)
+        [Universe.type0, n]
+    | _ :: _ as vs ->
+      List.map (fun v -> (v, n)) vs
+    | exception Not_found ->
+      (* Either an unbound template universe due to missing arguments, or a
+         global one appearing in the inductive arity. *)
+      [Universe.make u, n]
+  in
+  let u = List.map_append map u in
+  if List.is_empty u then
+    (* No constraints, fall in Prop *)
+    Sorts.prop
+  else
+    let fold accu (u, n) = Universe.sup accu (supern u n) in
+    Sorts.sort_of_univ (List.fold_left fold Universe.type0 u)
```

Comment:
```suggestion
  match u with
  | [] ->
    (* No constraints, fall in Prop *)
    Sorts.prop
  | (u,n) :: rest ->
    let fold accu (u, n) = Universe.sup accu (supern u n) in
    Sorts.sort_of_univ (List.fold_left fold (supern u n) rest)
```
I don't know if this fixes extlib but it is subtly different. eg
~~~coq
Inductive Box (A:Type) := box (_:A).
Inductive Box' (A:Type) := box' (_:Box A).
Check Box' True : Prop.
~~~
(I'll push this change)

Reply:
??? It does fix extlib but this seems very dangerous (and it goes against my mental model of template poly). Can't we use this to break soundness?

Reply:
Remember that while typing an inductive (ie inside Box') univ variables are not >= Set.

Reply:
Yes, but I am a bit surprised this is compositional. Anyways, maybe not as dangerous as it seems but clearly surprising.